File: screener/agents/momentum.py
"""Momentum agent — rewards strong recent gains with rising volume."""

from __future__ import annotations

from typing import Any, Dict

from .base import AgentResult, BaseAgent


class MomentumAgent(BaseAgent):
    name = "momentum"
    description = "Strong multi-timeframe gains with rising volume and healthy RSI"

    def evaluate(self, m: Dict[str, Any], context: Dict[str, Any] | None = None) -> AgentResult:
        reasons = []
        flags = []

        c7 = m.get("change_7d", 0.0)
        c20 = m.get("change_20d", 0.0)
        rsi = m.get("rsi_14", 50.0)
        v5 = m.get("vol_trend_5d", 0.0)
        v7 = m.get("vol_trend_7d", 0.0)
        macd = m.get("macd_hist", 0.0)
        pct_ma50 = m.get("pct_from_ma50", 0.0)

        # Score: gain × volume confirmation × healthy RSI
        gain_score = self.clip(c7 * 4, 0, 50)  # 0-50 for 0-12.5%+ gain
        vol_score = 0
        if v5 > 0 and v7 > 0:
            vol_score += 20
            reasons.append("vol rising 5d & 7d")
        elif v5 > 0 or v7 > 0:
            vol_score += 10
            reasons.append("vol partially rising")

        rsi_score = 0
        if 55 <= rsi <= 75:
            rsi_score = 15
            reasons.append(f"RSI healthy ({rsi:.0f})")
        elif rsi > 80:
            rsi_score = -10
            flags.append(f"RSI overextended ({rsi:.0f})")
        elif rsi < 45:
            rsi_score = -5

        macd_score = 8 if macd > 0 else -3
        if macd > 0:
            reasons.append("MACD bullish")

        ma_score = 7 if pct_ma50 > 0 else -3
        if pct_ma50 > 5:
            reasons.append(f"+{pct_ma50:.0f}% above 50-MA")

        if c7 > 8 and c20 > 0:
            reasons.append(f"7d +{c7:.1f}%, 20d +{c20:.1f}%")

        total = gain_score + vol_score + rsi_score + macd_score + ma_score
        score = self.clip(total)
        return AgentResult(self.name, score, self.tier_for(score), reasons, flags)

File: worker/src/engine/agents/momentum.ts
// Momentum agent — port of screener/agents/momentum.py.

import type { Metrics } from "../metrics";
import { type Agent, type AgentContext, type AgentResult, clip, fmt0, fmt1, tierFor } from "./base";

export const momentumAgent: Agent = {
  name: "momentum",
  description: "Strong multi-timeframe gains with rising volume and healthy RSI",

  evaluate(m: Metrics, _context: AgentContext): AgentResult {
    const reasons: string[] = [];
    const flags: string[] = [];

    const c7 = m.change_7d ?? 0;
    const c5 = m.change_5d ?? 0;
    const c20 = m.change_20d ?? 0;
    const rsi = m.rsi_14 ?? 50;
    const v5 = m.vol_trend_5d ?? 0;
    const v7 = m.vol_trend_7d ?? 0;
    const macd = m.macd_hist ?? 0;
    const pctMa50 = m.pct_from_ma50 ?? 0;
    void c5;

    const gainScore = clip(c7 * 4, 0, 50);

    let volScore = 0;
    if (v5 > 0 && v7 > 0) {
      volScore = 20;
      reasons.push("vol rising 5d & 7d");
    } else if (v5 > 0 || v7 > 0) {
      volScore = 10;
      reasons.push("vol partially rising");
    }

    let rsiScore = 0;
    if (rsi >= 55 && rsi <= 75) {
      rsiScore = 15;
      reasons.push(`RSI healthy (${fmt0(rsi)})`);
    } else if (rsi > 80) {
      rsiScore = -10;
      flags.push(`RSI overextended (${fmt0(rsi)})`);
    } else if (rsi < 45) {
      rsiScore = -5;
    }

    const macdScore = macd > 0 ? 8 : -3;
    if (macd > 0) reasons.push("MACD bullish");

    const maScore = pctMa50 > 0 ? 7 : -3;
    if (pctMa50 > 5) reasons.push(`+${fmt0(pctMa50)}% above 50-MA`);

    if (c7 > 8 && c20 > 0) reasons.push(`7d +${fmt1(c7)}%, 20d +${fmt1(c20)}%`);

    const score = clip(gainScore + volScore + rsiScore + macdScore + maScore);
    return { name: this.name, score, tier: tierFor(score), reasons, flags };
  },
};
